data/test/scenarios/wml_tests/ScenarioWML/EffectWML/test_remove_ability_by_filter.cfg

```diff
@@ -0,0 +1,132 @@
+#####
+# API(s) being tested: [effect]apply_to=remove_ability[filter_ability]
+##
+# Actions:
+# Two [chance_to_hit] abilities are added, setting the chance to 0 and 100 respectively.
+# The one setting the chance to 100 is removed, filtering by id and tag_name.
+# Have alice attack bob.
+##
+# Expected end state:
+# All strikes for both units miss.
+#####
+{GENERIC_UNIT_TEST "test_remove_ability_by_filter" (
+    [event]
+        name=start
+        [modify_unit]
+            [filter]
+            [/filter]
+            max_hitpoints=100
+            hitpoints=100
+            attacks_left=1
+        [/modify_unit]
+        [object]
+            silent=yes
+            [effect]
+                apply_to=new_ability
+                [abilities]
+                    [chance_to_hit]
+                        id=test_cth_0
+                        value=0
+                    [/chance_to_hit]
+                    [chance_to_hit]
+                        id=test_cth
+                        value=100
+                    [/chance_to_hit]
+                [/abilities]
+            [/effect]
+            [filter]
+                id=bob
+            [/filter]
+        [/object]
+        [object]
+            silent=yes
+            [effect]
+                apply_to=new_ability
+                [abilities]
+                    [chance_to_hit]
+                        id=test_cth_0
+                        value=0
+                    [/chance_to_hit]
+                    [chance_to_hit]
+                        id=test_cth
+                        value=100
+                    [/chance_to_hit]
+                [/abilities]
+            [/effect]
+            [filter]
+                id=alice
+            [/filter]
+        [/object]
+        [object]
+            [filter]
+                id=bob
+            [/filter]
+            silent=yes
+            [effect]
+                apply_to=remove_ability
+                [filter_ability]
+                    id=test_cth
+                    tag_name=chance_to_hit
+                [/filter_ability]
+            [/effect]
+        [/object]
+        [object]
+            [filter]
+                id=alice
+            [/filter]
+            silent=yes
+            [effect]
+                apply_to=remove_ability
+                [filter_ability]
+                    id=test_cth
+                    tag_name=chance_to_hit
+                [/filter_ability]
+            [/effect]
+        [/object]
+        [store_unit]
+            [filter]
+                id=alice
+            [/filter]
+            variable=a
+            kill=yes
+        [/store_unit]
+        [store_unit]
+            [filter]
+                id=bob
+            [/filter]
+            variable=b
+        [/store_unit]
+        [unstore_unit]
+            variable=a
+            find_vacant=yes
```

Comment:
These unit tests can use a `COMMON_KEEP` macro and `simulate_combat` to reduce the amount of boilerplate code, and so make it easier to see what's being tested. `simulate_combat` also means it can test the damage and number of blows rather than having to check Alice's remaining HP.

I think I posted code before without the other parts needed to make it work (probably missing the `COMMON_KEEP`), so here's a full working example:

```
#####
# API(s) being tested: [abilities][overwrite_specials][overwrite_filter] 
##
# Actions:
# Bob is given three [damage] abilities, one with value=5, other with add=11 and the last with multiply=2.
# The one with multiply=2 has overwrite_specials=both_sides and [overwrite_filter] to overwrite [damage]add=11 but not to overwrite [damage]value=5.
# Have bob attack alice.
##
# Expected end state:
# Bob's attack does 10 damage per hit, because the add=11 was overwritten but the value=5 and multiply=2 are still active.
#####
{COMMON_KEEP_A_B_UNIT_TEST "test_overwrite_specials_ability" (
    [event]
        name=start

        [object]
            silent=yes
            [effect]
                apply_to=new_ability
                [abilities]
                    [damage]
                        id=test_dmg
                        value=5
                    [/damage]
                    [attacks]
                        id=test_att
                        value=1
                    [/attacks]
                    [damage]
                        id=test_overwrite
                        add=11
                    [/damage]
                    [damage]
                        id=test_overwrite_specials
                        multiply=2
                    [/damage]
                    [overwrite_specials]
                        id=test_ability
                        [overwrite_filter]
                            tag_name=damage
                            type_value=add
                        [/overwrite_filter]
                    [/overwrite_specials]
                    [chance_to_hit]
                        value=100
                    [/chance_to_hit]
                [/abilities]
            [/effect]
            [filter]
                id=bob
            [/filter]
        [/object]

        [lua]
            code=<<
                local alice = wesnoth.units.find({id="alice"})[1]
                local bob = wesnoth.units.find({id="bob"})[1]

                -- Bob attacks Alice, doing it this way round ensures that it's melee rather than ranged combat.
                _, _, att_weapon, _ = wesnoth.simulate_combat(bob, alice)
                unit_test.assert_equal(att_weapon.num_blows, 1, "Bob’s weapon should only have 1 attack")
                unit_test.assert_equal(att_weapon.damage, 10, "Bob’s weapon should do 10 damage per hit")
            >>
        [/lua]

        {SUCCEED}
    [/event]
)}
```